Integrate common reachability service

Guardiola31337 opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->

Replaces our internal `NetworkStatusService` with the Common SDK's `ReachabilityInterface`.

### Changelog
<!--
Include changelog entry (e.g. Fixed an unexpected change in recenter button when resuming the app.).
See https://github.com/mapbox/navigation-sdks/blob/main/documentation/android-changelog-guidelines.md.
You can remove the changelog block and add a `skip changelog` label, when applicable.
 -->
```
<changelog>Navigation SDK now respects the `OfflineSwitch` and will not make resource request if the `setMapboxStackConnected` is set to `false`, it will use cached resources and immediately fallback to local generation of routes, even if connectivity is available.</changelog>
```

@kkaefer @mr1sunshine @tm243 Initial findings / questions

Currently, I'm getting 3 updates right at startup:

```
2021-06-21 18:39:11.734 26966-26966/com.mapbox.navigation.examples E/DEBUG: DEBUG ConnectivityHandler.status=REACHABLE_VIA_WI_FI
2021-06-21 18:39:12.110 26966-26966/com.mapbox.navigation.examples E/DEBUG: DEBUG ConnectivityHandler.status=REACHABLE_VIA_WI_FI
2021-06-21 18:39:12.113 26966-26966/com.mapbox.navigation.examples E/DEBUG: DEBUG ConnectivityHandler.status=REACHABLE_VIA_WI_FI
```

and when switching to `Airplane mode` still get the same status `REACHABLE_VIA_WIFI`

```
2021-06-21 18:40:15.333 26966-26966/com.mapbox.navigation.examples E/DEBUG: DEBUG ConnectivityHandler.status=REACHABLE_VIA_WI_FI
```

instead of `NOT_REACHABLE`.

- How's that suppose to work? Am I missing something or is this a bug?
- `reachability` takes a `hostname` which I believe is fine if we pass `null` here to get overall connectivity updates, is that right?
- Does that adding / removing an observer implies `start` / `stop` or should we also call these explicitly?

Guardiola31337 commented:
cc'ing @tm243 for visibility

LukasPaczos commented on libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ConnectivityHandler.kt:
```diff
@@ -0,0 +1,23 @@
+    fun getNetworkStatusChannel(): ReceiveChannel<Boolean> = networkStatusChannel
+
+    override fun run(status: NetworkStatus) {
+        Log.e("DEBUG", "DEBUG ConnectivityHandler.status=$status")
```
Use `Logger`.

Guardiola31337 replied:
This was only for debugging purposes. I wasn't planning to leave it out afterwards - don't want to spam out the `Logcat`, can do if you think otherwise.

LukasPaczos replied:
Ah, gotcha. I actually do think that it would be beneficial to log the connectivity state.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -181,6 +183,7 @@ class MapboxNavigation(
     private val tripSession: TripSession
     private val navigationSession: NavigationSession
     private val logger = LoggerProvider.logger
+    private val connectivityHandler: ConnectivityHandler = ConnectivityHandler()
```
As far as I can tell from only looking at the code, this handler instance is never notified, right?

The `ReachabilityService` creates another instance in place, so this one is never used.

LukasPaczos reviewed (commented):
Changelog update suggestion to highlight what this change actually means for developers:
```
<changelog>Navigation SDK now respects the `OfflineSwitch` and will not make resource request if the `setMapboxStackConnected` is set to `false`, it will use cached resources and immediately fallback to local generation of routes, even if connectivity is available.</changelog>
```

Guardiola31337 commented:
@LukasPaczos Addressed the comments. This is ready for another round of eyes from a technical point of view although I'm still able to reproduce the `REACHABLE_VIA_WIFI` status issue cc @tm243 @mr1sunshine

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -258,6 +260,8 @@ class MapboxNavigation(
      */
     val historyRecorder = MapboxHistoryRecorder(navigationOptions, logger)
 
+    private var reachabilityObserverId: Long? = null
```
```suggestion
    private val reachabilityObserverId: Long
```

The observer is added in the `init` block so we should be able to make it a value.

Guardiola31337 replied:
We're checking for `null` 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/ca0f9c4ab5729403ac09af42be30b60c49efdf3b/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L283

https://github.com/mapbox/mapbox-navigation-android/blob/ca0f9c4ab5729403ac09af42be30b60c49efdf3b/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L518

I think that's not an option.

LukasPaczos replied:
@Guardiola31337 this does the trick. Just a nit, not a blocker.
```diff
diff --git a/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt b/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
index 645700d7f..49307b137 100644
--- a/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
+++ b/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
@@ -260,7 +260,7 @@ class MapboxNavigation(
      */
     val historyRecorder = MapboxHistoryRecorder(navigationOptions, logger)
 
-    private var reachabilityObserverId: Long? = null
+    private val reachabilityObserverId: Long
 
     init {
         ThreadController.init()
@@ -280,11 +280,9 @@ class MapboxNavigation(
             MapboxModuleProvider.createModule(MapboxModuleType.NavigationRouter, ::paramsProvider),
             logger
         )
-        if (reachabilityObserverId == null) {
-            reachabilityObserverId = ReachabilityService.addReachabilityObserver(
-                connectivityHandler
-            )
-        }
+        reachabilityObserverId = ReachabilityService.addReachabilityObserver(
+            connectivityHandler
+        )
         directionsSession.registerRoutesObserver(navigationSession)
         val notification: TripNotification = MapboxModuleProvider
             .createModule(MapboxModuleType.NavigationTripNotification, ::paramsProvider)
@@ -515,10 +513,7 @@ class MapboxNavigation(
         MapboxNavigationTelemetry.unregisterListeners(this@MapboxNavigation)
         ThreadController.cancelAllNonUICoroutines()
         ThreadController.cancelAllUICoroutines()
-        ifNonNull(reachabilityObserverId) {
-            ReachabilityService.removeReachabilityObserver(it)
-            reachabilityObserverId = null
-        }
+        ReachabilityService.removeReachabilityObserver(reachabilityObserverId)
     }
 
     /**
```

Guardiola31337 replied:
Ohh yeah, I wanted to prevent any bad usage when creating / destroying `MapboxNavigation`. Making sure there's always only 1 active `ReachabilityObserver`.

LukasPaczos commented on libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ConnectivityHandler.kt:
```diff
@@ -0,0 +1,29 @@
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ReceiveChannel
+
+class ConnectivityHandler(private val logger: Logger) : ReachabilityChanged {
```
Missing a test for this class.

LukasPaczos reviewed (commented):
LGTM from the integration perspective. 2 small comments below and waiting for the resolution of the issue you mentioned.

tm243 commented:
Looks good to me.
The issue with switching to Airplane mode has been addressed and discussed with @Guardiola31337
The 3 NetworkStatus=REACHABLE_VIA_WI_FI at startup are normal.

LukasPaczos commented:
@tm243 @Guardiola31337 I can also reproduce the reachability interface reporting `REACHABLE_VIA_WI_FI` after I switch to airplane mode. Why is that expected?

LukasPaczos commented:
Ah, I now see this is not expected and there's a PR resolving the issue. I don't think we can merge this PR before that issue is fixed and available in a Common SDK version/patch, that would be a regression.

Guardiola31337 commented:
@LukasPaczos About to push a workaround for that. Currently testing all scenarios locally.

Guardiola31337 commented:
@LukasPaczos @tm243 this should be good now for a final review 🚀 

cc @zugaldia
